pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// current revision. updateSet is the representation of the set at the updateRevision. currentRevision is the name of
+// the current revision. updateRevision is the name of the update revision. ordinal is the ordinal of the Pod. If the
+// returned error is nil, the returned Pod is valid.
+func newVersionedStatefulSetPod(currentSet *apps.StatefulSet, updateSet *apps.StatefulSet, currentRevision string, updateRevision string, ordinal int) *v1.Pod {
+	if (currentSet.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		ordinal < int(currentSet.Status.CurrentReplicas)) ||
+		(currentSet.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType &&
+			ordinal < int(currentSet.Spec.UpdateStrategy.Partition.Ordinal)) {
+		pod := newStatefulSetPod(currentSet, ordinal)
+		setPodRevision(pod, currentRevision)
+		return pod
+	}
+	pod := newStatefulSetPod(updateSet, ordinal)
+	setPodRevision(pod, updateRevision)
+	return pod
+}
+
+// getPatch returns a strategic merge patch that can be applied to restore a StatefulSet to a
+// previous version. If the returned error is nil the patch is valid.
+func getPatch(set *apps.StatefulSet) ([]byte, error) {
+	str, err := runtime.Encode(patchCodec, set)
+	if err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	json.Unmarshal([]byte(str), &raw)
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+	spec := raw["spec"].(map[string]interface{})
+	template := spec["template"].(map[string]interface{})
+	specCopy["template"] = template
+	template["$patch"] = "replace"
+	objCopy["spec"] = specCopy
+	patch, err := json.Marshal(objCopy)
+	return patch, err
+}
+
+// newRevision creates a new ControllerRevision containing a patch that can convert a StatefulSet to an equivalent state to set.
+// The Revision of the returned ControllerRevision is set to Revision. If the returned error is nil. the returned
+// ControllerRevision is valid.
+func newRevision(set *apps.StatefulSet, revision int64) (*apps.ControllerRevision, error) {
+	patch, err := getPatch(set)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return history.NewControllerRevision(set,
+		controllerKind,
+		selector,
+		runtime.RawExtension{Raw: patch},
+		revision)
+}
+
+// applyRevision returns a new StatefulSet constructed by restoring the state in revision to set. If the returned error
+// is nil, the returned StatefulSet is valid.
+func applyRevision(set *apps.StatefulSet, revision *apps.ControllerRevision) (*apps.StatefulSet, error) {
+	obj, err := scheme.Scheme.DeepCopy(set)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*apps.StatefulSet)
+	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), revision.Data.Raw, clone)
```

Comment:
I added a check under strategicpatch
If you mean check that the clone casts correctly, we generally do not do that when deep copying from using Scheme.